File: recalld/pipeline/vault.py
from __future__ import annotations

from datetime import date
from typing import Optional
from urllib.parse import quote

import httpx

from recalld.pipeline.align import LabelledTurn
from recalld.pipeline.postprocess import PostProcessResult


PREVIEW_MAX_CHARS = 1200


def _render_session_note_body(
    session_date: date,
    category: str,
    speakers: list[str],
    result: Optional[PostProcessResult],
    turns: list[LabelledTurn],
) -> str:
    post_processing_status = "failed" if result is None else "ok"
    speakers_yaml = "[" + ", ".join(speakers) + "]"
    date_str = session_date.isoformat()

    transcript_lines = "\n".join(f"> **{t.speaker}:** {t.text}" for t in turns)

    if result is None:
        body = "_Post-processing failed. Transcript preserved below._\n"
        focus_section = ""
    else:
        focus_items = "\n".join(f"- [ ] {p}" for p in result.focus_points)
        body = f"{result.summary}\n\n[Full transcript ↓](#transcript)\n"
        focus_section = f"\n## Focus\n\n{focus_items}\n"

    return f"""---
date: {date_str}
category: {category}
speakers: {speakers_yaml}
post_processing: {post_processing_status}
---

## Summary

{body}{focus_section}
## Transcript

> [!note]- Full transcript
{transcript_lines}
"""


def render_session_note(
    session_date: date,
    category: str,
    speakers: list[str],
    result: Optional[PostProcessResult],
    turns: list[LabelledTurn],
) -> str:
    return _render_session_note_body(session_date, category, speakers, result, turns)


def _strip_frontmatter(markdown: str) -> str:
    if markdown.startswith("---\n"):
        parts = markdown.split("\n---\n", 1)
        if len(parts) == 2:
            return parts[1].lstrip()
    return markdown.strip()


def _truncate_preview(markdown: str, max_chars: int = PREVIEW_MAX_CHARS) -> str:
    text = markdown.strip()
    if len(text) <= max_chars:
        return text

    cutoff = text.rfind("\n\n", 0, max_chars)
    if cutoff < max_chars // 2:
        cutoff = text.rfind("\n", 0, max_chars)
    if cutoff == -1:
        cutoff = max_chars

    return text[:cutoff].rstrip() + "\n\n..."


def render_session_note_preview(
    session_date: date,
    category: str,
    speakers: list[str],
    result: Optional[PostProcessResult],
    turns: list[LabelledTurn],
    max_chars: int = PREVIEW_MAX_CHARS,
) -> str:
    """Return a shortened markdown excerpt suitable for HTML preview rendering."""
    note = _render_session_note_body(session_date, category, speakers, result, turns)
    return _truncate_preview(_strip_frontmatter(note), max_chars=max_chars)


def render_focus_section(session_date: date, focus_points: list[str]) -> str:
    items = "\n".join(f"- [ ] {p}" for p in focus_points)
    return f"\n## {session_date.isoformat()}\n\n{items}\n"


class VaultWriteError(Exception):
    pass


class VaultWriter:
    def __init__(self, api_url: str, api_key: str) -> None:
        self.api_url = api_url.rstrip("/")
        self.api_key = api_key

    def _headers(self) -> dict[str, str]:
        headers: dict[str, str] = {}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        return headers

    async def write_note(self, vault_path: str, filename: str, content: str) -> None:
        encoded = quote(f"{vault_path}/{filename}", safe="/")
        url = f"{self.api_url}/vault/{encoded}"
        async with httpx.AsyncClient(verify=False, timeout=15.0) as client:
            resp = await client.put(url, content=content.encode(), headers={
                **self._headers(),
                "Content-Type": "text/markdown",
            })
            if resp.status_code >= 400:
                raise VaultWriteError(f"Obsidian API error {resp.status_code}: {resp.text}")

    async def patch_heading(self, vault_path: str, heading: str, content: str) -> None:
        encoded = quote(vault_path, safe="/")
        heading_encoded = quote(heading, safe="")
        url = f"{self.api_url}/vault/{encoded}/heading/{heading_encoded}"
        async with httpx.AsyncClient(verify=False, timeout=15.0) as client:
            resp = await client.put(url, content=content.encode(), headers={
                **self._headers(),
                "Content-Type": "text/markdown",
            })
            if resp.status_code >= 400:
                raise VaultWriteError(f"Obsidian API error {resp.status_code}: {resp.text}")

    async def append_to_heading(self, vault_path: str, heading: str, content: str) -> None:
        encoded = quote(vault_path, safe="/")
        heading_encoded = quote(heading, safe="")
        url = f"{self.api_url}/vault/{encoded}/heading/{heading_encoded}"
        async with httpx.AsyncClient(verify=False, timeout=15.0) as client:
            resp = await client.post(url, content=content.encode(), headers={
                **self._headers(),
                "Content-Type": "text/markdown",
            })
            if resp.status_code >= 400:
                raise VaultWriteError(f"Obsidian API error {resp.status_code}: {resp.text}")

    async def append_to_note(self, vault_path: str, content: str) -> None:
        encoded = quote(vault_path, safe="/")
        url = f"{self.api_url}/vault/{encoded}"
        async with httpx.AsyncClient(verify=False, timeout=15.0) as client:
            resp = await client.patch(url, content=content.encode(), headers={
                **self._headers(),
                "Content-Type": "text/markdown",
            })
            if resp.status_code >= 400:
                raise VaultWriteError(f"Obsidian API error {resp.status_code}: {resp.text}")

    async def read_note(self, vault_path: str) -> str:
        encoded = quote(vault_path, safe="/")
        url = f"{self.api_url}/vault/{encoded}"
        async with httpx.AsyncClient(verify=False, timeout=15.0) as client:
            resp = await client.get(url, headers=self._headers())
            if resp.status_code == 404:
                return ""
            if resp.status_code >= 400:
                raise VaultWriteError(f"Obsidian API error {resp.status_code}: {resp.text}")
            return resp.text

    async def list_directory(self, vault_path: str) -> list[str]:
        encoded = quote(vault_path, safe="/")
        url = f"{self.api_url}/vault/{encoded}/"
        async with httpx.AsyncClient(verify=False, timeout=15.0) as client:
            resp = await client.get(url, headers=self._headers())
            if resp.status_code == 404:
                return []
            if resp.status_code >= 400:
                raise VaultWriteError(f"Obsidian API error {resp.status_code}: {resp.text}")
            data = resp.json()
            files = data.get("files", [])
            return [str(item) for item in files if isinstance(item, str)]

    async def note_exists(self, vault_path: str) -> bool:
        encoded = quote(vault_path, safe="/")
        url = f"{self.api_url}/vault/{encoded}"
        async with httpx.AsyncClient(verify=False, timeout=5.0) as client:
            resp = await client.get(url, headers=self._headers())
            return resp.status_code == 200

    async def open_note(self, vault_path: str) -> None:
        encoded = quote(vault_path, safe="/")
        url = f"{self.api_url}/open/{encoded}"
        async with httpx.AsyncClient(verify=False, timeout=5.0) as client:
            resp = await client.post(url, headers=self._headers())
            if resp.status_code >= 400:
                raise VaultWriteError(f"Obsidian API error {resp.status_code}: {resp.text}")

File: vault.go
package main

import (
	"bytes"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

const previewMaxChars = 1200

type PostProcessResult struct {
	Summary     string   `json:"summary"`
	FocusPoints []string `json:"focus_points"`
	RawResponse string   `json:"raw_response,omitempty"`
	Strategy    string   `json:"strategy"`
	TopicCount  int      `json:"topic_count"`
}

func renderSessionNoteBody(sessionDate time.Time, category string, speakers []string, result *PostProcessResult, turns []LabelledTurn) string {
	postProcessingStatus := "failed"
	if result != nil {
		postProcessingStatus = "ok"
	}
	speakersYAML := "[" + strings.Join(speakers, ", ") + "]"
	dateStr := sessionDate.Format("2006-01-02")

	var transcriptLines []string
	for _, turn := range turns {
		transcriptLines = append(transcriptLines, fmt.Sprintf("> **%s:** %s", turn.Speaker, turn.Text))
	}

	body := "_Post-processing failed. Transcript preserved below._\n"
	focusSection := ""
	if result != nil {
		var focusItems []string
		for _, p := range result.FocusPoints {
			focusItems = append(focusItems, "- [ ] "+p)
		}
		body = result.Summary + "\n\n[Full transcript ↓](#transcript)\n"
		if len(focusItems) > 0 {
			focusSection = "\n## Focus\n\n" + strings.Join(focusItems, "\n") + "\n"
		} else {
			focusSection = "\n## Focus\n\n"
		}
	}

	return fmt.Sprintf(`---
date: %s
category: %s
speakers: %s
post_processing: %s
---

## Summary

%s%s## Transcript

> [!note]- Full transcript
%s
`, dateStr, category, speakersYAML, postProcessingStatus, body, focusSection, strings.Join(transcriptLines, "\n"))
}

func RenderSessionNote(sessionDate time.Time, category string, speakers []string, result *PostProcessResult, turns []LabelledTurn) string {
	return renderSessionNoteBody(sessionDate, category, speakers, result, turns)
}

func stripFrontmatter(markdown string) string {
	if strings.HasPrefix(markdown, "---\n") {
		parts := strings.SplitN(markdown, "\n---\n", 2)
		if len(parts) == 2 {
			return strings.TrimLeft(parts[1], "\n")
		}
	}
	return strings.TrimSpace(markdown)
}

func truncatePreview(markdown string, maxChars int) string {
	text := strings.TrimSpace(markdown)
	if len(text) <= maxChars {
		return text
	}

	cutoff := strings.LastIndex(text[:maxChars], "\n\n")
	if cutoff < maxChars/2 {
		cutoff = strings.LastIndex(text[:maxChars], "\n")
	}
	if cutoff == -1 {
		cutoff = maxChars
	}
	return strings.TrimRight(text[:cutoff], "\n ") + "\n\n..."
}

func RenderSessionNotePreview(sessionDate time.Time, category string, speakers []string, result *PostProcessResult, turns []LabelledTurn, maxChars int) string {
	if maxChars <= 0 {
		maxChars = previewMaxChars
	}
	note := renderSessionNoteBody(sessionDate, category, speakers, result, turns)
	return truncatePreview(stripFrontmatter(note), maxChars)
}

func RenderFocusSection(sessionDate time.Time, focusPoints []string) string {
	items := make([]string, 0, len(focusPoints))
	for _, p := range focusPoints {
		items = append(items, "- [ ] "+p)
	}
	return fmt.Sprintf("\n## %s\n\n%s\n", sessionDate.Format("2006-01-02"), strings.Join(items, "\n"))
}

type VaultWriter struct {
	APIURL string
	APIKey string
	Client *http.Client
}

func (w VaultWriter) client() *http.Client {
	if w.Client != nil {
		return w.Client
	}
	return &http.Client{Timeout: 15 * time.Second}
}

func (w VaultWriter) headers() http.Header {
	h := make(http.Header)
	if w.APIKey != "" {
		h.Set("Authorization", "Bearer "+w.APIKey)
	}
	return h
}

func (w VaultWriter) WriteNote(vaultPath, filename, content string) error {
	encoded := url.PathEscape(vaultPath + "/" + filename)
	encoded = strings.ReplaceAll(encoded, "%2F", "/")
	req, err := http.NewRequest(http.MethodPost, strings.TrimRight(w.APIURL, "/")+"/vault/"+encoded, bytes.NewReader([]byte(content)))
	if err != nil {
		return err
	}
	req.Header = w.headers()
	req.Header.Set("Content-Type", "text/markdown")
	resp, err := w.client().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return fmt.Errorf("obsidian api error %d", resp.StatusCode)
	}
	return nil
}

func (w VaultWriter) AppendToNote(vaultPath, content string) error {
	encoded := url.PathEscape(vaultPath)
	encoded = strings.ReplaceAll(encoded, "%2F", "/")
	req, err := http.NewRequest(http.MethodPatch, strings.TrimRight(w.APIURL, "/")+"/vault/"+encoded, bytes.NewReader([]byte(content)))
	if err != nil {
		return err
	}
	req.Header = w.headers()
	req.Header.Set("Content-Type", "text/markdown")
	resp, err := w.client().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return fmt.Errorf("obsidian api error %d", resp.StatusCode)
	}
	return nil
}

func (w VaultWriter) NoteExists(vaultPath string) (bool, error) {
	encoded := url.PathEscape(vaultPath)
	encoded = strings.ReplaceAll(encoded, "%2F", "/")
	req, err := http.NewRequest(http.MethodGet, strings.TrimRight(w.APIURL, "/")+"/vault/"+encoded, nil)
	if err != nil {
		return false, err
	}
	req.Header = w.headers()
	resp, err := w.client().Do(req)
	if err != nil {
		return false, err
	}
	defer resp.Body.Close()
	return resp.StatusCode == http.StatusOK, nil
}

func (w VaultWriter) OpenNote(vaultPath string) error {
	encoded := url.PathEscape(vaultPath)
	encoded = strings.ReplaceAll(encoded, "%2F", "/")
	req, err := http.NewRequest(http.MethodPost, strings.TrimRight(w.APIURL, "/")+"/open/"+encoded, nil)
	if err != nil {
		return err
	}
	req.Header = w.headers()
	resp, err := w.client().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return fmt.Errorf("obsidian api error %d", resp.StatusCode)
	}
	return nil
}
